### PR TITLE
Initial Prometheus support

### PR DESCRIPTION
**Pull request by johscheuer**
Problem: Currently the F5 CTLR is a blackbox for the user (the only way to gather information is parsing the logs).

Solution: Add Prometheus metrics for the most important statistics e.g. How many services managed, how many (and where) error occurred, how many nodes are used, e.g.. Also sometimes the Python process inside the container dies e.g. when a malformed configuration file is passed for this purpose we can expose a health endpoint that checks if the Python process is still running. This allows to configure a Kubernetes health check that restarts the container if needed.

Any further ideas which metrics could be useful (these can be extended in a later PR anyways)?

TODOs:
- [X] Add Prometheus metric for monitored nodes
- [X] Add Prometheus metric for managed services
- [X] Add Prometheus metric for malformed configurations
- [X] Clean up Prometheus metrics for deleted services
- [X] Add health endpoint
- [X] Document new cli argument
- [X] Add documentation for the endpoints

### CODE REVIEW

**Comment by amudukutore** [comment]
@johscheuer - this is a great idea and was on the F5 team's radar as well. Someone from the F5 team will get back to you in the next week or two to brainstorm on what metrics/status data we can capture, and figure out mechanisms to export these.

**Comment by johscheuer** [comment]
@amudukutore any further information about a short meeting?

**Comment by johscheuer** [comment]
The current output looks like the following (I excluded the default go metrics provided by the Prometheus go client):

```
curl -s localhost:8080/metrics | grep bigip  
# HELP bigip_monitored_nodes Total count of monitored nodes by the BigIP k8s CTLR
# TYPE bigip_monitored_nodes gauge
bigip_monitored_nodes{nodeselector="kubernetes.io/role=node"} 3
# HELP bigip_monitored_services Total count of monitored services by the BigIP k8s CTLR
# TYPE bigip_monitored_services gauge
bigip_monitored_services{name="prometheus",namespace="ns1",status="parse-error"} 0
bigip_monitored_services{name="prometheus",namespace="ns1",status="port-not-found"} 0
bigip_monitored_services{name="prometheus",namespace="ns1",status="service-not-found"} 0
bigip_monitored_services{name="prometheus",namespace="ns1",status="success"} 1
bigip_monitored_services{name="prometheus",namespace="monitoring",status="parse-error"} 0
bigip_monitored_services{name="prometheus",namespace="monitoring",status="port-not-found"} 0
bigip_monitored_services{name="prometheus",namespace="monitoring",status="service-not-found"} 0
bigip_monitored_services{name="prometheus",namespace="monitoring",status="success"} 1
```

**Comment by amudukutore** [comment]
@johscheuer - our architect is out until later this week. I will schedule a short meeting once he's back (probably for Friday morning Denver time (UTC-7).

**Comment by amudukutore** [comment]
@johscheuer - sent you an invite by email for a meeting tomorrow (10 AM Denver time).

**Comment by f5gary** [comment]
Hi @johscheuer. I am going to be looking at this PR in the very near future, but I see there are some conflicts. Would you mind addressing those conflicts and update the PR? Thank you.

**Comment by johscheuer** [comment]
@f5gary thanks! I rebased the PR. 

I will try to resolve the open todos in the next week (most is only documentation work).

**Comment by johscheuer** [comment]
I just stumbled across this:

Currently the AppManager does the following steps:

1.  Check if service port exists
2. Check if service exist

Wouldn't it be smarter to check if there service exist and than check if the service contains the requested port?

File: pkg/health/health.go [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package health
 
 import (
+	"log"
```
Comment (f5gary) [inline pkg/health/health.go]: Please use our vlogger package for consistency. (import log "github.com/F5Networks/k8s-bigip-ctlr/pkg/vlogger")

File: pkg/health/health.go [hunk 2]
```diff
@@ -13,6 +14,7 @@ type HealthChecker struct {
 //TODO: add health check if Kubernetes API is still reachable
 func (hc HealthChecker) HealthCheckHandler() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log.Printf("Health check called, checking for subprocess with ID: %d\n", hc.SubPID)
```
Comment (f5gary) [inline pkg/health/health.go]: This should be a debug-level message (log.Debugf())

File: pkg/health/health.go [hunk 3]
```diff
@@ -21,6 +23,7 @@ func (hc HealthChecker) HealthCheckHandler() http.Handler {
 				w.Write([]byte("Ok"))
 				return
 			}
+			log.Println(err)
```
Comment (f5gary) [inline pkg/health/health.go]: This should be an error-level message (log.Error())

File: pkg/appmanager/appManager.go [hunk 4]
```diff
@@ -959,15 +955,22 @@ func (appMgr *Manager) syncConfigMaps(
 			stats.vsUpdated += updated
 		}
 
+		log.Errorf("Deactivated: %v, %v", deactivated, cm.ObjectMeta.Name)
```
Comment (f5gary) [inline pkg/appmanager/appManager.go]: This shouldn't be an error message, probably info.

File: pkg/health/health.go (hunk 1 above)
Comment (f5gary) [reply pkg/health/health.go]: Nevermind, I see this was changed in a later commit.

File: pkg/health/health.go (hunk 3 above)
Comment (f5gary) [reply pkg/health/health.go]: Nevermind, I see this was changed in a later commit.

File: pkg/appmanager/appManager.go (hunk 4 above)
Comment (f5gary) [reply pkg/appmanager/appManager.go]: Nevermind, I see this was changed in a later commit.

File: docs/README.rst [hunk 5]
```diff
@@ -97,6 +97,9 @@ General
 |                       |         |          |                   | to verify the BIG-IP                    |                |
 |                       |         |          |                   | configuration.                          |                |
 +-----------------------+---------+----------+-------------------+-----------------------------------------+----------------+
+| http-listen-address   | string  | Optional | "0.0.0.0:8080"    | Address to serve http based informations|                |
+|                       |         |          |                   | e.g. (`/metrics` and ``health`)         |                |
```
Comment (f5gary) [inline docs/README.rst]: The back-ticks around 'health' do not match. This causes a warning during the "make docs" step.

**Comment by johscheuer** [comment]
I updated the parts you mentioned. If there is anything missing let me know! :)

File: cmd/k8s-bigip-ctlr/main.go [hunk 6]
```diff
@@ -26,12 +27,15 @@ import (
...
 	"github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager"
+	"github.com/F5Networks/k8s-bigip-ctlr/pkg/health"
 	"github.com/F5Networks/k8s-bigip-ctlr/pkg/pollers"
+	bigIPPrometheus "github.com/F5Networks/k8s-bigip-ctlr/pkg/prometheus"
 	"github.com/F5Networks/k8s-bigip-ctlr/pkg/vxlan"
 	"github.com/F5Networks/k8s-bigip-ctlr/pkg/writer"
 
 	log "github.com/F5Networks/k8s-bigip-ctlr/pkg/vlogger"
 	clog "github.com/F5Networks/k8s-bigip-ctlr/pkg/vlogger/console"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
```
Comment (russokj) [inline cmd/k8s-bigip-ctlr/main.go]: If not already done, we need to get legal to sign off on any additional 3rd party packages.

Comment (f5gary) [reply cmd/k8s-bigip-ctlr/main.go]: I'll take care of that. I'd like to get this merged, but it's a WIP and I don't know why.

Comment (johscheuer) [reply cmd/k8s-bigip-ctlr/main.go]: @f5gary I resolved the WIP status (sorry that I forgot the change that).

Comment (f5gary) [reply cmd/k8s-bigip-ctlr/main.go]: Thank you @johscheuer.

**Comment by johscheuer** [comment]
Rebased - resolved the merge conflict

File: cmd/k8s-bigip-ctlr/main.go (hunk 6 above)
Comment (f5gary) [reply cmd/k8s-bigip-ctlr/main.go]: FYI, I have verified we do not have any new 3rd party packages.

**Comment by mdditt2000** [comment]
Updated Jira for PM track CONTCNTR-1210

**Pull request merged**